About docs

oligriffiths opened the pull request:
Added an about doc to explain how broccoli works and the basic concepts.
Comment away with suggestions, it's important we make this guide simple from a nomenclature and concepts POV.

Also added support for a markdown table of contents.

`yarn serve` to test.

stefanpenner commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+title: What is Broccoli?
+---
+
+Broccoli is a simple Javascript build tool, that allows you to configure your build pipeline in Javascript (like you're
```
JavaSciprt

stefanpenner commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+these all operate on different levels to Broccoli.
+
+Broccoli works at the file-system level, it provides a Javascript API to wrap other node tools like
+[Babel](https://babeljs.io/), [Rollup](https://rollupjs.org/) or [Node-Sass](https://github.com/sass/node-sass), that
```
webpack, browserify, and many more

stefanpenner commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+
+## Thinking in Broccoli
+
+Broccoli at its heart is very simple. It's really just a set of connected file transformations, performed by plugins,
```
Maybe something like:

It is really just a set of directories, connected together by a plugin system. This plugin systems allow behavior between two directories to be described. Broccoli then ensures plugins are called at the appropriate time, to ensure the build is succesfull.

stefanpenner commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+that manipulate files from one or multiple inputs, into a single output. Each plugin is responsible for processing files
+passed to it in input directories, and writing files to its output directory.
+
+Broccoli handles managing the file-system state, and simply takes input files and passes them as inputs to plugins, 
```
broccoli doesn't really care about files, it's simply connects directories and behavior together to achieve what the user wants. Custom behavior is provided by plugins, which are  responsible to read from their input directories, and write to their single output directory.

stefanpenner commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+and takes the output from each plugin and passes it as an input to the next one. The order in which these operations 
+happen is determined by the build file, a `Brocfile.js` that you're going to learn how to create.
+
+You can think of Broccoli much like a simple programming language, where the output of a function can be passed as the
```
maybe `s/broccoli/broccoli-plugins`

stefanpenner commented on src/docs/about.md:
````diff
@@ -0,0 +1,186 @@
+]);
+```
+
+In the above, working from inner to outer, a `src` directory is passed to the `babel()` plugin (which will convert our
````
Rather than saying "from inner to outer" might it be clearer to have a number list of steps which sequentializes the pipeline in question? I at-least find "inner to outer" to be hard to follow.

stefanpenner commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+
+### Plugins
+
+Plugins are what a build pipeline developer will interact with most. Plugins are what do the actual work of transforming
```
We spoke about this in todays call, I will review again once you had time to iterate on this section.

stefanpenner commented:
cc @mike-north could I also get your eyes on this?

stefanpenner commented on src/docs/about.md:
```diff
@@ -0,0 +1,208 @@
+title: What is Broccoli?
+---
+
+Broccoli is a simple JavaScript build tool, that allows you to configure your build pipeline in Javascript (like
```
`Javascript` -> `JavaScript`

stefanpenner commented on src/docs/about.md:
````diff
@@ -0,0 +1,208 @@
+}
+```
+
+Basic plugins are super simple, anything you can do in node, you can do in the `build()` method. A plugin can
````
I think we should avoid saying phrases like `super simple`. Rather lets show how they are "super simple". For example, "a broccoli-plugin has only 1 purpose to transform the files from this.inputPaths directories to their appropriate output in this.outputPath directory when its build function is invoked."

stefanpenner commented on src/docs/about.md:
```diff
@@ -0,0 +1,208 @@
+directly to its output directory, and optionally allows Broccoli to be notified when files within the input 
+directory change and trigger a rebuild.
+
+### Nodes
```
Do you think we can get away without talking about nodes? I think for the most part, just thinking about broccoli-plugins being connected by broccoli in a graph is sufficient. Yes that implies broccoli-plugins are nodes, but I worry raising nodes and pointing to graph-theory provides less value than cost.

oligriffiths replied:
Yeah I think that's fair.

oligriffiths commented on src/docs/about.md:
```diff
@@ -0,0 +1,208 @@
+
+Broccoli works at the file-system level, it provides a Javascript API to wrap other node tools like
+[Babel](https://babeljs.io/), [Rollup](https://rollupjs.org/) or [Node-Sass](https://github.com/sass/node-sass), 
+[Webpack](https://webpack.js.org/), [Browserify](http://browserify.org/), and many more. These tools operate on a
```
Correct, suggestions welcome

mike-north commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+
+## Thinking in Broccoli
+
+Broccoli at its heart is very simple. It is really just a set of directories, connected together by a plugin
```
```suggestion
Broccoli consists of a set of directories, connected together by plugins, which describe how files are moved or transformed at each step of the build process.
```

mike-north commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+## Thinking in Broccoli
+
+Broccoli at its heart is very simple. It is really just a set of directories, connected together by a plugin
+system. This plugin systems allow behavior between input and output directories to be described. Broccoli then
```
```suggestion
Broccoli
```

mike-north commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+
+Broccoli at its heart is very simple. It is really just a set of directories, connected together by a plugin
+system. This plugin systems allow behavior between input and output directories to be described. Broccoli then
+ensures plugins are called at the appropriate time, to ensure the build is successful, and writes the result to
```
```suggestion
ensures plugins are called in the prescribed order, and writes the result to
```

mike-north commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+---
+
+Broccoli is a simple JavaScript build tool, that allows you to configure your build pipeline in JavaScript (like
+you're already used to writing), and handles setting up the filesystem state for each transformation that's 
```
I would not necessarily assume that readers understand the distinction between an asset pipeline and a module bundler.

You could say something like

> Broccoli is a build tool, that through allowing developers to compose modular plugins to form an asset pipeline, quickly and reliably combines and transforms source files into deployable static assets.

mike-north commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+
+Broccoli doesn't really care about files, it simply takes source directories and passes them as inputs to 
+plugins, creates an output directory for the plugin to write to, and passes that output directory it as an input
+to the next plugin.
```
You're taking a lot of time to try and describe something that might be easier to show via a diagram

```
app                                       dist
 └─ src						         	   └─ prod.build-20190206.js
     ├─ index.ts      -->  MyPlugin -->
     └─ other.ts

```

mike-north commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+### Plugins
+
+Plugins are what a build pipeline developer will interact with most. Plugins are what do the actual work of
+transforming input files into output files. The API of a plugin is pretty simple, all that's required is creating
```
you may want to be careful of input/output as terminology, since they could be taken to mean pipeline imports or plugin imports.

```suggestion
transforming files at each step of build process. The API of a plugin is pretty simple, all that's required is creating
```

mike-north commented on src/docs/about.md:
````diff
@@ -0,0 +1,186 @@
+}
+```
+
+A broccoli-plugin has only 1 purpose to transform the files from `this.inputPaths` directories to their output
````
```suggestion
A broccoli-plugin has only one purpose: to transform the files from `this.inputPaths` directories to their output
```

mike-north commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+state for each node, creating an `outputPath` for each node to write to.
+
+You may often encounter the term "tree" when reading plugin READMEs or in tutorials, just remember a tree is a
+connected set of plugins/nodes.
```
One of the important concepts to get across is that trees can be 

- a string
- something a plugin emits

mike-north commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+
+## Building
+
+Broccoli build pipelines are defined using a `Brocfile.js` file in the root of the project. This `js` file sets up
```
> Broccoli build pipelines are defined using a `Brocfile.js` file in the root of the project.
feels like you've said this a few times already

mike-north commented on src/docs/about.md:
```diff
@@ -0,0 +1,186 @@
+## Building
+
+Broccoli build pipelines are defined using a `Brocfile.js` file in the root of the project. This `js` file sets up
+the build graph, connecting source directories via plugins, and exports a single node whose output will be written
```
Is it important that consumers of broccoli know about a "build graph"? This sounds like it might be an implementation detail

stefanpenner commented on src/content/about.md:
```diff
@@ -0,0 +1,239 @@
+
+Broccoli works by managing of a set of directories, connected together by plugins, which describe how files 
+are moved or transformed at each step of the build process. Broccoli ensures plugins are called in the prescribed
+order, and writes the result to a `target` directory. Each plugin is responsible for processing files passed to 
```
It might be to low level, but an interesting fact, is if you use the same plugin instance twice, it doesn't result in duplicate work, as plugins auto memoize on a per rebuild basis.

feel free to disregard if this is the wrong place to mention this.

oligriffiths replied:
I feel like that's too low level

stefanpenner commented:
Left some minor notes, otherwise LGTM pending @mike-north's sign-off.

Q: should we describe how broccoli-builder can be used in another system? Or is that out of scope for this piece?

lisaychuang commented on src/content/about.md:
```diff
@@ -0,0 +1,239 @@
+to the next plugin.
+
+Broccoli is configured with a file in the root of your project called `Brocfile.js`. This file defines the build
+pipeline for your application, and is written in plan old JavaScript. The order in which operations happen is
```
```suggestion
pipeline for your application, and is written in plain old JavaScript. The order in which operations happen is
```

lisaychuang commented on src/content/about.md:
````diff
@@ -0,0 +1,239 @@
+```
+
+A broccoli-plugin has only one purpose to transform the files from `this.inputPaths` directories to their output
+directory in this.outputPath directory when its `build()` function is invoked. Anything you can do in node, 
````
```suggestion
directory in `this.outputPath` directory when its `build()` function is invoked. Anything you can do in node, 
```

lisaychuang commented on src/content/about.md:
```diff
@@ -0,0 +1,239 @@
+you can do in the `build()` method. A plugin can receive one or multiple inputs, and these are available in the 
+`this.inputPaths` array in the order they are provided. `this.inputPaths` contains paths to directories, that are
+the `outputPath` of previous plugins. Each `inputPath` contains files that you can manipulate and write to
+`this.outputPath`. Broccoli will handle the state of these directories and passing them between plugins.
```
```suggestion
`this.outputPath`. Broccoli will handle the state of these directories and take responsibility for passing them between plugins.
```

lisaychuang commented on src/content/about.md:
```diff
@@ -0,0 +1,239 @@
+There is a special case where a `string` is passed as an input to a plugin. When parsing your build pipeline, 
+Broccoli will automatically convert a string input into a
+[source plugin](https://github.com/broccolijs/broccoli-source). This plugin basically connects its input directory
+directly to its output directory, and also allows Broccoli to `watch` and be notified when files within the input
```
```suggestion
to its output directory, and also allows Broccoli to `watch` and be notified when files within the input
```

oligriffiths commented:
@stefanpenner seems like an advanced doc to me

lisaychuang commented on src/content/about.md:
```diff
@@ -0,0 +1,239 @@
+Working from the target directory (the final output) back up to the source directories resembles a tree. Think 
+of a piece of Broccoli and you should have a mental model of what a build pipeline looks like.
+
+Broccoli constructs this tree of connected nodes in memory as it parses the `Brocfile`, and sets up the filesystem 
```
Having read this for the first time, I don't know what a `Brocfile` is by this point.  Maybe call if a configuration file, since you will talk about Brocfile again in ##Building section below.

lisaychuang commented on src/content/about.md:
```diff
@@ -0,0 +1,239 @@
+plugin, and finally resolve all the symlinks and write the files from the final node into the destination build 
+directory.
+
+Confused? Here's an example:
```
```suggestion
Here's an example:
```

lisaychuang commented on src/content/about.md:
```diff
@@ -0,0 +1,239 @@
+
+Thus `module.exports` contains a node that references the two input nodes, and an output path that will contain the
+contents of `dir1` and `dir2` when the `build` command is run. The two input nodes reference two source
+directories, `dir1` and `dir2`.
```
Suggest to reverse these 2 sentences.

oligriffiths commented:
@stefanpenner @mike-north I think we're good to go, addressed the feedback and I think clarified a few other sections, thanks for the reviews.

@lisaychuang Thank you for the review, comments addressed and alterations made.

lisaychuang commented:
Glad my review was helpful, thanks @oligriffiths !

stefanpenner commented on src/content/about.md:
```diff
@@ -0,0 +1,255 @@
+### Trees
+
+The build pipeline is series of connected plugins, from one or more input directories to a single target directory.
+Working from the target directory (the final output) back up to the source directories resembles a tree. Think of a
```
What you have here may be fine, but let me go into more detail so you can make the call if the docs are still appropriate or not.

---

The input structure isn't always tree, for example:

```js
const a = new Foo('foo');
const b = new Bar(a);

merge([a,b]);
```

Although the output of ever plugin is tree like.
Also internally broccoli does cycle breaking, and only processes the plugins as if they were a tree, but connects them as a directed graph.

oligriffiths replied:
Sure, but I think from a mental model, it's still useful to think of it as a tree.
```
  a
/  \
|   b
\  /
 c
```
The above is more of a shrub than a tree

![](https://www.naturehills.com/media/wysiwyg/blog/Privacy_Fence_LG.jpg)

If you feel strongly about this, please say so, but I think the overall mental model is useful to think of like a tree.

stefanpenner replied:
Nah, I'm onboard. I just wanted to make sure this detail doesn't detract from the mental model. I do agree with you, I think it's fine. Saying "broccoli trees are actually directed graphs" likely just confuses the story, for little gain.

stefanpenner commented on src/content/about.md:
```diff
@@ -0,0 +1,255 @@
+You may often encounter the term "tree" when reading plugin READMEs or in tutorials, just remember a tree is a
+connected set of plugins.
+
+## Building
```
should we mention that broccoli guarantees to only be building 1 plugin at a time per pipeline.

oligriffiths replied:
updated

stefanpenner commented on src/content/getting-started.md:
````diff
@@ -40,7 +40,9 @@ if required, however installing within the project allows for `yarn` script comm
 
 ```shell
-$ echo "module.exports = 'src'" > Brocfile.js && mkdir src
+$ echo "module.exports = 'src'" > Brocfile.js
````
Should this be updated, considering the recent improvements re:

* esm
* function export

mike-north commented on src/content/about.md:
```diff
@@ -91,60 +91,70 @@ app                                                       [target]
 
+Note: [ES Modules](https://nodejs.org/api/esm.html) syntax is supported by Broccoli in the current master. Prior
+to this, standard CommonJs syntax with `require` and `module.exports` should be used.
+
```
out of curiosity, does this basically just mean you're providing default exports now?

oligriffiths replied:
The standard practice will be to use `export default` from the Brocfile. Is that what you mean?
